engine/src/main/java/org/destinationsol/files/HullConfigManager.java

```diff
@@ -53,21 +57,37 @@ public HullConfigManager(ItemManager itemManager, AbilityCommonConfigs abilityCo
         configToNameMap = new HashMap<>();
     }
 
-    private static Vector2 readVector2(JsonValue jsonValue, String name, Vector2 defaultValue) {
-        String string = jsonValue.getString(name, null);
-        return (string == null)
-                ? defaultValue
-                : SolMath.readV2(string);
+    private static @Nullable Vector2 readOptionalVector2(@NotNull JsonValue jsonValue, @Nullable String name, @Nullable Vector2 defaultValue) {
+        Optional<String> string = Optional.ofNullable(jsonValue.getString(name, null));
+
+        return string.map(SolMath::readV2).orElse(defaultValue);
     }
 
-    private static Engine.Config readEngineConfig(String engineName, ItemManager itemManager) {
-        if (engineName == null)
-            return null;
+    /**
+     * Reads and parses a position field from given {@link JsonValue}.
+     *
+     * The position field is expected to be of type string, in the form of {@code "0.5 0.5"}. This would return a {@link Vector2} with values 0.5f, 0.5f.
+     *
+     * @param jsonValue Json whose position you want to get
+     * @param superComponent Descriptive name of parameter whose position you are trying to get. Will be used as part of exception message in case the position is not found in the Json
+     * @return {@code Vector2} representation of the position
+     */
+    private static @NotNull Vector2 readPosition(@NotNull JsonValue jsonValue, @Nullable String superComponent) {
+        Optional<String> string = Optional.ofNullable(jsonValue.getString("position", null));
+
+        return string
+                .map(SolMath::readV2)
+                .orElseThrow(() -> new SolDescriptiveException("Some of your " + superComponent + "s is missing a required \"position\" argument, or it is malformed."));
```

Comment:
Not sure about the syntax of the exception.
"Some of your gunslots is missing a.."

Reply:
Also, here is the problem: You are telling the user "some" gunSlot. Certainly better than a simple json parse exception, but which gunslot? I would prefer we throw an exception here, that propagates further up. This will throw `jsonParse`, the caller will throw `gunSlot -> jsonParseException`, and the caller will throw `shipName -> gunSlot -> jsonParseException` which will finally be displayed. Does this make sense?

Reply:
Not quite